vunk-parser/src/expr.rs

```diff
@@ -0,0 +1,585 @@
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct DeclArg<'src> {
+    pub name: Option<&'src str>,
+    pub ty: DeclType<'src>,
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct DefRhs<'src> {
+    pub args: Vec<(&'src str, DeclType<'src>)>,
+    pub expr: Box<Expr<'src>>,
+}
+
+type ParserInput<'tokens, 'src> =
+    chumsky::input::SpannedInput<Token<'src>, SimpleSpan, &'tokens [(Token<'src>, SimpleSpan)]>;
+
+impl Expr<'_> {
+    pub fn parser<'tokens, 'src: 'tokens>() -> impl Parser<
+        'tokens,
+        ParserInput<'tokens, 'src>,
+        Expr<'src>,
+        chumsky::extra::Err<Rich<'tokens, Token<'src>, SimpleSpan>>,
+    > + Clone {
+        chumsky::recursive::recursive(|expr| {
+            // TODO: Support floats
+            // let float_parser = select! {
+            // };
+
+            let list_parser = {
+                let left_br = just(Token::Op("[")).to(());
+                let right_br = just(Token::Op("]")).to(());
+
+                expr.clone().delimited_by(left_br, right_br)
+            };
+
+            let unary_parser = {
+                let unary_op_parser = {
+                    let binary = just(Token::Op("~")).to(UnaryOp::BinaryNot);
+                    let logical = just(Token::Op("!")).to(UnaryOp::LogicalNot);
+
+                    binary.or(logical)
+                };
+
+                unary_op_parser
+                    .then(expr.clone())
+                    .map(|(op, expr)| Expr::Unary {
+                        op,
+                        expr: Box::new(expr),
+                    })
+            };
+
+            let binary_parser = {
+                let binary_op_parser = just(Token::Op("+"))
+                    .to(BinaryOp::Add)
+                    .or(just(Token::Op("-")).to(BinaryOp::Sub))
+                    .or(just(Token::Op("*")).to(BinaryOp::Mul))
+                    .or(just(Token::Op("/")).to(BinaryOp::Div))
+                    .or(just(Token::Op("%")).to(BinaryOp::Rem))
+                    .or(just(Token::Op("==")).to(BinaryOp::Eq))
+                    .or(just(Token::Op("!=")).to(BinaryOp::NotEq))
+                    .or(just(Token::Op("<")).to(BinaryOp::Less))
+                    .or(just(Token::Op("<=")).to(BinaryOp::LessEq))
+                    .or(just(Token::Op(">")).to(BinaryOp::More))
+                    .or(just(Token::Op(">=")).to(BinaryOp::MoreEq))
+                    .or(just(Token::Op("&")).to(BinaryOp::BitAnd))
+                    .or(just(Token::Op("&&")).to(BinaryOp::LogicalAnd))
+                    .or(just(Token::Op("|")).to(BinaryOp::BitOr))
+                    .or(just(Token::Op("||")).to(BinaryOp::LogicalOr))
+                    .or(just(Token::Op("^")).to(BinaryOp::BitXor))
+                    .or(just(Token::Op("++")).to(BinaryOp::Join));
+
+                expr.clone()
+                    .then(binary_op_parser)
+                    .then(expr.clone())
```

Comment:
This looks like it might be left-recursive. If you have issues here, I recommend a [precedence climbing](https://github.com/zesterer/chumsky/blob/legacy/tutorial.md#parsing-unary-operators) approach instead.